context.go

```diff
@@ -6,53 +6,142 @@ import (
 	"flag"
 	"fmt"
 	"strings"
+	"time"
 )
 
-// Context is a type that is passed through to
-// each Handler action in a cli application. Context
+// Context is an interface that is passed through to
+// each Handler action in a cli application. It
 // can be used to retrieve context-specific args and
 // parsed command-line options.
-type Context struct {
-	context.Context
-	App           *App
-	Command       *Command
+type Context interface {
+	// Functions to be able to manage Context type
+	App() *App
+	SetCommand(command *Command)
+	Command() *Command
+	Context() context.Context
+	ParentContext() Context
+	SetShellComplete(shellComplete bool)
```

Comment:
I'm going to pick on this as an example, although I'm sure it's not the only one.

When creating an interface that we expose, we need to be conservative about what methods we're exposing publicly. Having a setter method here seems to be an artifact of how the code happened to have been written, but I can't think of a scenario in which a user would ever be calling this function.

In this case, since the one place where this is called is right after we create a new context, we could find a way to get this information into the context constructor and remove it from the interface. On the more general note though, I'd like to see this as small and immutable as we can possibly make it.

Reply:
I like where you are going with this. Having done this first step, it allowed me to gain a perspective over the grand scheme of things. I now will incorporate the feedback here slowly as we try to lessen the public face of the API.